### ISSUE TITLE
`archivePackages()`: Account for all `type` variants by default

### ISSUE THREAD

**Issue by pat-s** [eddelbuettel/drat#90]
Over the past weeks I've written a more sophisticated version of `archivePackages()` which accounts for (all available) binaries (including all existing mac binary versions).

It is already in use at https://github.com/mlr3learners/mlr3learners.drat.
Currently, 21 packages live in that repo with binaries for R >= 3.6.
Binaries are build once a month or for every release.
After that, the adjusted `drat::archivePackages()` is called in a CI run to clean up.

Over the last weeks my plan was to make a PR including these changes. Now I've just seen #89 was merged two days ago and resulted in a new CRAN release.
While #89 is welcome and adds support for binaries in general and accounts for the new binary naming scheme on macOS, the diff between drat upstream and my version became substantial now.
I assume you might not like another substantial change/refactoring to the code base in such a short time, that's why I am first opening this issue rather than putting time in to rebase against drat upstream.
I also do not want to trigger a long discussion about this.
I would appreciate a short and concise decision whether the following points sound attractive or not.

- In contrast to the current `drat::archivePackages()`, my version archives for all possible `type` inputs automatically.
- Also, no custom `contrib.url2()` is needed.

If these points are attractive to you, I am happy to clean up and open a PR.
If not, I will continue maintaining my own drat fork.
Until now I stayed close to CRAN drat and avoided side changes, with the aim of opening a PR.
However, if my proposal here gets declined I am considering adding more substantial changes to my fork and branch off from CRAN drat completely, maybe even renaming and refactoring the whole package (of course with the required acknowledgements).

My current approach lives in https://github.com/pat-s/drat/tree/upd-archive-pkgs.

**Comment by eddelbuettel** [comment]
`archivePackages()` has always been a user-contributed function; I have taken contributions before and there is no reason why we can't do that again, preferably if the existing interface are maintained.  If you clean code that improves the current situation, I am sure folks who use `archivePackages()` would be happy to review, starting with @FelixErnst who made the most recent change.  Small and incremental changes are always preferred of large and more disruptive 'waterfall' changes.

**Comment by FelixErnst** [comment]
I just compared the two branches (https://github.com/eddelbuettel/drat/compare/master...pat-s:upd-archive-pkgs) and I have some follow-up questions.

- are those the correct branches to compare?
- the essence of the changes you propose are "vectorization" on `archivePackage` and `pruneRepo` an the `type` argument, am I correct?

I like the general idea of the changes, but I think just slapping a `for` and `lapply` on the `type` argument leaves room for improvement (I mean `for (type in type)` ... ). Also fixing the `type` to certain values, without enforcing them from the start (with `match.arg`), will potentially result in typos and calls to the function leading to complicated result states, especially, if you have a `stop` in a `lapply`.

In addition, `pruneRepo` does not returns, which version information it has found for which `type`, thus `pruneRepo` probably returns non unique rows. In you example this work, because `archivePackages` calls `pruneRepo` with a single type.

Regarding `contrib.url2()`, you should checkout the issue leading to the PR to understand, why it is there (I found @jangorecki's comment interesting, that it is a feature, which might sorely missing from base R). So, I would be interested, if you try to archive a `mac.binary`/ `mac.binary.el-capitan` from R `3.4`. I don't see a way, in which `contrib.url` returns the correct path for that.

So I would like to see the changes introduced, but it has to take care of all the use cases we can think of.

**Comment by pat-s** [comment]
Thanks Felix.

> leaves room for improvement 

There is definitely room for improvement. I opened this issue to get early feedback to figure out whether I'll contribute here or not (this relates mainly to sticking to other parts of the package that I would change substantially if I would focus in my fork).

> the essence of the changes you propose are "vectorization" on archivePackage and pruneRepo an the type argument, am I correct?

Yes.

>  So, I would be interested, if you try to archive a mac.binary/ mac.binary.el-capitan from R 3.4. I don't see a way, in which contrib.url returns the correct path for that.

So far I did not know about this issue as I did not apply my branch on R 3.4 binaries.
If issues occur there that cannot be handled at all, `contrib.url2()` has its relevance, no question that. I did not read all the bits and pieces of your PR to understand why everything was added :)
Regarding the issue: Did you consider a patch to R-core?

> So I would like to see the changes introduced, but it has to take care of all the use cases we can think of.

I'd argue it can grow over time and not everything has to be handled in one PR.
For now I'd focus on vectorization while taking into account your `contrib.url2()`.

So in summary, I'll open a clean PR that includes my change in the next days? Happy to take your review then.

Other points discussed in this PR should be followed up in dedicated issues.

**Comment by FelixErnst** [comment]
So I had a try for myself, since it was hard to describe what I thought might be possible without trying.

In my opinion, two places need to be vectorized:
- the part in `pruneRepo`, which gathers the information on the different packages and version for each type
- `contrib.url2` to accept multiple type information

`archivePackage` and the action part in `pruneRepo` don't need to changed, since the are just about subsetting and one action per row.

To make the structure of the code a bit clearer, I split `getRepoInfo` of from `pruneRepo`. This also makes the calls in `archivePackage` to `pruneRepo`/`getRepoInfo` more descriptive.

Have a look at the changes [here](https://github.com/eddelbuettel/drat/compare/master...FelixErnst:vectorize-archive-prune). They are completly untested and are just about illustrating the point, that a small restructuring with everything in mind, might be better.

**Comment by FelixErnst** [comment]
In addition, the result of `contrib.url2` must still be matched to the `type`. Maybe through naming the result...

Looking forward to the PR in the next week!

**Comment by jangorecki** [comment]
Motivation for my `contrib.url` was the fact I wanted to produce paths for R 4.0.0 and R 4.1.0 from a single R session. I build my repo using linux machine and I publish r-release and r-devel for package sources and windows binaries. Base R `contrib.url` is giving me paths only for currently running version of R, and my "publish machine" did not even have r-devel. Not related to macosx at all.

**Comment by eddelbuettel** [comment]
The initially instinct is to infer OS version and release (if any) from the host, but that is not always desirable or feasible as one may host on one OS with binaries coming from other OSs or build service.  Ideally we should be able to deal with binaries in a self-describing manner but that is both harder and subject to almost pernanent change so unsure if we can in all cases -- maybe explicit arguments are easier. But I just don't know as it is simply not my primary use case.

**Comment by FelixErnst** [comment]
> Motivation for my `contrib.url` was the fact I wanted to produce paths for R 4.0.0 and R 4.1.0 from a single R session. I build my repo using linux machine and I publish r-release and r-devel for package sources and windows binaries. Base R `contrib.url` is giving me paths only for currently running version of R, and my "publish machine" did not even have r-devel. Not related to macosx at all.

Nonetheless, it is the same issue, since the path generation is the purpose of `contrib.url`. I think base R needs it for internal functions anyway, so I think, its quite safe to regard this function as a cornerstone and build on it as needed.

The addition of a version argument would be a nice addition to base R, but is there a use case within base R, which would require it to be added? I am not sure... and I guess, if this question would be clear yes, it would have been added long ago.

**Comment by eddelbuettel** [comment]
Duncan Murdoch also often (rightly) responds on _e.g._ r-package-devel by (and I paraphrase) "if it works well in a package why does it have to be in base R?" which is a fair point as well.
